### ISSUE TITLE
Server onError Handling

### ISSUE THREAD

**Issue by benjchristensen** [ReactiveX/RxNetty#90]
The server does not have default error handling when `Observable<Void>` calls onError.

Right now I have to manually do it or errors are swallowed:

``` java
                    return testEndpoint(request, response).onErrorFlatMap(error -> {
                        return writeError(request, response, "Unknown error: " + error.getMessage());
                    });
```

**Comment by NiteshKant** [comment]
I guess if the method HttpServer.withErrorResponseGenerator() was public (which was the original intent :) )

https://github.com/Netflix/RxNetty/blob/master/rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/HttpServer.java#L43

it would suffice, rite? 

One can define a global catch all error response generator using the above.

**Comment by benjchristensen** [comment]
Yes I should be able to override it, but there should be a default like servlet containers that returns a status 500. Right now it just silently fails and swallows errors.

**Comment by NiteshKant** [comment]
Actually, there is a default error generator that sets the status to 500. If you create a server like:

```
RxNetty.createHttpServer(9999, new RequestHandler<ByteBuf, ByteBuf>() {
                                              @Override
                                              public Observable<Void> handle(
                                                      HttpServerRequest<ByteBuf> request,
                                                      HttpServerResponse<ByteBuf> response) {
                                                  return Observable.error(new UnsupportedOperationException());
                                              }
                                          }).startAndWait();
```

You will get an error response as:

```
curl -v "http://localhost:9999"
* About to connect() to localhost port 9999 (#0)
*   Trying ::1...
* connected
* Connected to localhost (::1) port 9999 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.24.0 (x86_64-apple-darwin12.0) libcurl/7.24.0 OpenSSL/0.9.8y zlib/1.2.5
> Host: localhost:9999
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
* Closing connection #0
```

However, there is a caveat. The error response generator is called if and only if, a response (partial or full) is not already written on that channel. So, if you instead use the following request handler:

```
                                          new RequestHandler<ByteBuf, ByteBuf>() {
                                              @Override
                                              public Observable<Void> handle(
                                                      HttpServerRequest<ByteBuf> request,
                                                      HttpServerResponse<ByteBuf> response) {
                                                  response.writeString("hello");
                                                  return Observable.error(new UnsupportedOperationException());
                                              }
                                          }
```

which writes a content before throwing an error, the error response generator will not be called.

Of course there is an issue here that in this case, the error is swallowed. Is this the case you have in your example?

**Comment by NiteshKant** [comment]
@benjchristensen as part of pull request #95 I have made the aforementioned [method](https://github.com/Netflix/RxNetty/blob/master/rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/HttpServer.java#L43) public. Does that suffice? Can I close this issue?

**Comment by benjchristensen** [comment]
Should we make the default include a message body, like Apache HTTP and Tomcat (and every other server?) does? Right now this will show a blank response unless verbosely looking at the response code.

**Comment by NiteshKant** [comment]
Yeah sure I can add a global error response. Should we be returning an HTML or json?

**Comment by benjchristensen** [comment]
I don't have a strong preference for that. It's primarily for humans so they don't fumble around with empty responses trying to understand what's going on and then go implement their own handler.

**Comment by fiunchinho** [comment]
Nice!
Is there an easy way to overwrite this default error response generator (say for JSON response) other than overwriting the HttpRxServerProvider ? Seems overkill.

Also, @NiteshKant comitted a main() method in the DefaultErrorResponseGenerator class. I guess you were using it to test it and you forgot it was there.

**Comment by NiteshKant** [comment]
@fiunchinho you can use [`HttpServer.withErrorResponseGenerator()`](https://github.com/ReactiveX/RxNetty/blob/0.4.x/rxnetty/src/main/java/io/reactivex/netty/protocol/http/server/HttpServer.java#L82) to specify a custom error response generator.

The change is more than a year old :) and yeah I think I removed the `main()` method after that.

**Comment by fiunchinho** [comment]
Actually, that class is missing in the 0.5 tag for RxNetty.
I was using Karyon and it seems there is no easy way to call the method HttpServer.withErrorResponseGenerator() from karyon. Karyon (even Karyon 3) depends on RxNetty 0.4, which seems odd.

**Comment by NiteshKant** [comment]
@fiunchinho [0.5.x](https://github.com/ReactiveX/RxNetty/wiki/0.5.x-FAQs) is still not stable (snapshot status) that is the reason why karyon would not be using 0.5.x just yet. 

`ErrorResponseGenerator` isn't available in 0.5.x by design, as you can achieve the same results by using an `.onErrorResumeNext()` operator, something like:

``` java
HttpServer.newServer()
                 .start((req, resp) ->
                            resp.writeString(just("Hello World!"))
.onErrorResumeNext(resp.setStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR)
                                          .writeString(Observable.just("Error!")))
                           );
```
